ibtk/include/ibtk/HierarchyAveragedDataManager.h

```diff
@@ -0,0 +1,293 @@
+    /*!
+     * The constructor for class HierarchyAveragedDataManager sets some default values and determines data centering.
+     * The expected period and number of snapshots must be set in the input database. This class assumes the snapshots
+     * will be taken at equidistant points along the periodic interval.
+     *
+     * The input database is searched for the following keys:
+     *  - 't_start' : Double that represents the beginning of the period
+     *  - 't_end' : Double that represents the end of the period. For cases where the period is zero, t_end should be
+     * set to t_start.
+     *  - 'threshold' : Double to determine whether a given time point is at a periodic steady state.
+     *  - 'num_snapshots' : Integer to determine the number of snapshots taken, equally sampled between t_start and
+     * t_end.
+     *  - 'enable_logging' : Bool used to determine whether to print convergence information to the log file.
+     *  - 'output_data' : Bool used to determine whether to write visualization files for the mean and deviations.
+     *  - 'dir_dump_name' : String used to determine which folder to write visualization files to.
+     *  - 'refine_type' : String to determine which refine operator will be used by default.
+     */
+    HierarchyAveragedDataManager(std::string object_name,
+                                 SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > var,
+                                 SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db,
+                                 SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > hierarchy,
+                                 SAMRAI::tbox::Pointer<SAMRAI::hier::GridGeometry<NDIM> > grid_geom,
+                                 bool register_for_restart = true);
+
+    /*!
+     * The constructor for class HierarchyAveragedDataManager sets some default values and determines data centering. In
+     * this constructor, the times at which snapshots are taken are set by arguments.
+     *
+     * The input database is searched for the following keys:
+     *  - 'enable_logging' : Bool used to determine whether to print convergence information to the log file.
+     *  - 'output_data' : Bool used to determine whether to write visualization files for the mean and deviations.
+     *  - 'dir_dump_name' : String used to determine which folder to write visualization files to. Defaults to "".
+     *  - 'refine_type' : String to determine which refine operator will be used by default. Defaults to
+     * "CONSERVATIVE_LINEAR_REFINE".
+     */
+    HierarchyAveragedDataManager(std::string object_name,
+                                 SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > var,
+                                 SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db,
+                                 SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > hierarchy,
+                                 std::set<double> snapshot_time_points,
```

Comment:
FYI: `std::set` technically cannot be used with `double` since `double`s are not totally ordered due to `NaN`s. This code (same with `std::map`) will break in complicated ways if a `NaN` gets in.

Reply:
Are you suggesting that I rewrite this to use `std::vector` and keep it sorted myself? Or just to check for `NaN` on construction?

Reply:
I'm not sure what the best solution but I think checking for NaNs suffices.